utils/tps/src/main.rs

```diff
@@ -18,39 +37,178 @@ struct Args {
 	/// Total number of expected transactions
 	#[arg(long)]
 	num: usize,
+
+	/// Whether to subscribe to blocks from genesis or not.
+	/// For zombienet tests, this should be set to true.
+	/// When deploying tps in more long-living networks, set this to false.
+	#[arg(short, long)]
+	genesis: bool,
+
+	/// Whether to monitor relay-chain, or para-chain finality.
+	/// If set to true, tps will subscribe to CandidateIncluded events on the relaychain node,
+	/// and scrape Balances Transfer events concurrently with a collator node RPC client.
+	#[arg(short, long)]
+	para_finality: bool,
+
+	/// Default parablock time set to 12s for sync-backing.
+	/// This should be set to 6.0s for async-backing.
+	#[arg(short, long, default_value_t = 12)]
+	default_parablock_time: u64
 }
 
-pub async fn calc_tps(api: &Api, n: usize) -> Result<(), Error> {
+async fn calc_parablock_tps(
+	para_api: &Api,
+	mut rx: Receiver<H256>,
+	default_parablock_time: u64,
+) -> Result<(), Error> {
 	let storage_timestamp_storage_addr = runtime::storage().timestamp().now();
+	let mut trx_in_parablock = 0;
+	while let Some(para_head) = rx.recv().await {
+		info!("TPS Counter ===> Received ParaHead: {:?}", para_head);
+		let parablock = para_api.blocks().at(Some(para_head)).await?;
+		let parabody = parablock.body().await?;
+		let parablock_number = parablock.number();
+		let previous_parablock_number = parablock_number - 1;
+		let maybe_previous_parablock_hash =
+			para_api.rpc().block_hash(Some(previous_parablock_number.into())).await?;
+
+		// Need to handle the case where we cannot get the previous parablock timestamp
+		let parablock_time = match maybe_previous_parablock_hash {
+			Some(hash) => {
+				let parablock_timestamp = para_api
+					.storage()
+					.fetch(&storage_timestamp_storage_addr, Some(para_head))
+					.await?
+					.unwrap();
+				let previous_parablock_timestamp = para_api
+					.storage()
+					.fetch(&storage_timestamp_storage_addr, Some(hash))
+					.await?
+					.unwrap();
+				let time_diff = parablock_timestamp - previous_parablock_timestamp;
+				info!("TPS Counter ===> Parablock time estimated at: {:?}", time_diff);
+				time_diff
+			},
+			// Assume default if unable to get the previous parablock from parablock number
+			None => {
+				warn!(
+					"TPS Counter ===> Assuming default parablock time of: {:?}",
+					default_parablock_time
+				);
+				Duration::as_secs_f64(&Duration::new(default_parablock_time, 0)) as u64
+			},
+		};
```

Comment:
The `unwrap()`s here have to be fixed - just noticed an error during testing.

Reply:
what was the error?

Reply:
Panic due to unwrap on `None` value.

Reply:
I believe you actually need something like this:
```
				let parablock_timestamp = para_api
					.storage()
-					.fetch(&storage_timestamp_storage_addr, Some(para_head))
+					.fetch(&storage_timestamp_storage_addr, Some(hash))
					.await?
					.unwrap();
				let previous_parablock_timestamp = para_api
					.storage()
-					.fetch(&storage_timestamp_storage_addr, Some(hash))
+					.fetch(&storage_timestamp_storage_addr, Some(previous_hash))
					.await?
					.unwrap();
```

Reply:
fixed here: https://github.com/paritytech/polkadot-stps/pull/43/commits/74dca75fe69417cf8d3730271f5112add58cae1b